Use createImageBitmap when available

OmarShehata opened the pull request:
This PR adds support for using `createImageBitmap` to move image decoding off the main thread. This doesn't make loading any particular image faster, but it greatly improves frame stutters by avoiding the expensive blocking image decode that happens on texture upload. It might also improve loading scene time since images can be decoded in parallel.

This builds off the other PR https://github.com/AnalyticalGraphicsInc/cesium/pull/6625. The problem there was that images created with `createImageBitmap` can be flipped on texture upload only when passing `ImageBitmapOptions` to the `createImageBitmap` was not supported. This is because it's faster to flip the image during decode. So I had to add a way to flip an image when fetching it. 

@mramato to make this easier to review, I recommend reading each commit separately:

* https://github.com/AnalyticalGraphicsInc/cesium/commit/96cbc181262b0cacba60bee3894dc221c237f693 adds the new feature detection functions needed for supporting this. The new spec show how they're expected to be used.
* https://github.com/AnalyticalGraphicsInc/cesium/commit/b772d527599c787968ba2d5b3e652811b0e3b82f this is the meat of this PR. This adds `flipImage` on `Resource`, and uses `createImageBitmap` on all image requests, as well as in `loadImageFromTypedArray`. It also kicks off the support check in the Viewer constructor. 
* https://github.com/AnalyticalGraphicsInc/cesium/commit/dca9ed1accb55fcd753552dbbc69185bf6e2d0b9 adds tests to verify that `createImageBitmap` behaves how we expect it to. Specifically, it tests that it's possible to flip the image either on fetch or on texture upload as expected, and reads back the pixels since this is the only way to know if the operation truly worked. 
* Finally, https://github.com/AnalyticalGraphicsInc/cesium/commit/824ceb12e6a38208c44609a034365c78cc55296d updates all the ImageryProvider specs to check for `ImageBitmap` instead of `Image` when it's supported.

### Why it's necessary to expose `flipY` on `fetchImage`

We can't make decisions about whether to flip inside `Resource` because whether we flip or not depends on what we're going to do with this image. All textures in CesiumJS are flipped by default during texture upload. This seems to be because imagery expects this orientation.

Models with textures will disable this flipping to get the correct orientation. This means that there's no one correct orientation. If we always flip, models will be upside down. If we never flip, imagery will be upside down. I spent some time trying to see if I could get all imagery to be flipped in the shader so we can just simply never have to worry about flipping during texture upload but I couldn't get it to work correctly with the projection.

### Tests & Results

To compare before/after, simply add `return false` to the first line of `FeatureDetection.supportsCreateImageBitmap`.

[This bike model](https://sketchfab.com/3d-models/racing-bicycle-223901fe94de4d55b6324e7d7e273469) is a good test, since it has huge textures that severely block the main thread when loading in. 

If the glTF model is in `SampleData/models/Bike` you can use this [local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=pVeNbts2EH4VNgUaeXNoO0uLNk6KJU7SZEjbwHZbrDBQ0NLZJkKRGknZcYu8+46iZFGOOwxbYMTS3X33f0d6yTRZcliBJqdEwooMwPA8pZ8LWrQfF68DJS3jEvR+m/yYyImdWA1/5WDsEGQC+r1KgBwTq3NoezaTPGWWK4nkGRMG2sQzLE9BoKoN3ZOnzMAtW4O+4/E9OrOFmuVCmFgDyPPc2qda56Bi9OEJbqFS2I3gcqbO1cM22cQgwYWz2w8DAmIX1o1MeMyseiIh2ZLPi8ivQWRbtieSlH9zoaawg27u14FXTuVjq49fhcASi2Us03aMScR6lbX6IxecyQtmgc60Sm+Mev2q24v2D7uH3YPuq4PD7rh3dNzt4ufrfqHN+pLTWKj43pXaRSxoqHvz/A/yca41yJ2I2l+XUOSWGopXrxKZMfaVVkIU3VewqK/yKGMxDFgKmg02Mn2vtQZREIJnRvGkTsZlRaKfPtyMv43uri+HlzuQkk0FjLmwCC1y3fA6dfFtfMo0T7nlSzCUJUlUWvI5cBl/J+ws+lEXMdfimOxT2sHPiKWZACwO6xRKTeec30PHK54jcD+sfswEHJNet+tpj626+B4RFzmhGImx2o1kcvbATR3+AGuAT0z+5uP/2g/RUxbfz7XKZTJQQukA5l7p6PZsfPluePZnIxcxk0tmtkpIPbVfi31XKr1isQ3VJjBjubCfmcghWnGZqBXSljzG8XoAMXRz0iY92m0FilY8sQvU4S1gz3HssS8F8ZfASoBYAJ8v7Dbk2lO3MGUneMHKVPHdb7A2Ov3DBslnJPKwZ6enpEtevKjMF+8tEvSB822mc2PzdNNLZQFLcr8WdorrtLnKRqUQZSbDrVNkq9Uw4P52CFURkU7tfSX+SLAH4WdKrMrcPJRvukxgVEbY8Wpb/d3gqcJl5yI9CNSFpsumDpuLyblw66H3shsU1E/nUFmMRs6R7U6WTQnKLK0WIKNipqgGlqzvcBK5gRa1jjHLZbGovUQjax5T/H/PrOYPdcOONZNmpnRqKDBjPyhtF5+ysbrCdk2uNNYuCg7JYNjcGriAOe4uE3Vpt01ev6FvXrcJPgczvBmp/7L3/t/u+zf7b4dcyvEgz9OvOEIXHDe8jF21MKpGTJ0OGTiWICy3iuiqcNjTWAySGwwUt5TVa0etJKCZFFwO8prJRGzfRUZ1ai6XbrC9ULRjH/nTbWIDXdSAvZFZbs98PwSNsYQU1bWqK41z5Ennbc4GFHhs/8yl8ToDent5Nf528fHLh2bBvX+N+1LUlMizBLPhyx6wHltUYfr0Cru6dhsavewOAiWACjVHTr8CbnRUsC0blQq8B+HaedaMOuC6e57NtexXhMfqwQ/ur64XDjfW/KnuyudnbOqOGtQ4yjAMzAFLeG62pDE7FoIzw4/kEU3x4OCZWJ+v7xSXNnoyte3dRry+Ntk5p1ErXF7+7GCJr3Rtf4GLa4hk3AVREec2KOM2XvwU0nu5LY9bZe7mRuMyPdrRHOWhIJS6P7PRjgiuvZd3zu7QKYtKv9vel7Y3EayaiS2hZeOdVRdyv8XCdti0zV5778TYtYC3nvA7TzPcgO4+E+FVxgJeZRBlOtMcL8eWxsbUBp+7cRwsnBtJmzz3IZXvYccm3KCW9TGRSoZLp76bHMTuNnJMNCSNpVQQVwtuKxg240mncvgk4UvCk9PJ3tZvlskeiQUzBjnuZ8SIf4fJ3tuTDso3YEIVGf24BI3+OZFF7+2tJ1JKTzr4+hRllRJThkbKnNWcIB/IdbdmUr5WWpryjYwhwhdnG1N9/w0
) to see it loading while rotating, which makes the frame jank obvious.

Image bitmap on the left:

![compare_optimize](https://user-images.githubusercontent.com/1711126/52967366-de95e100-3377-11e9-9afe-94619b29807f.gif)

Image bitmap on top:

![image_bitmap_bike](https://user-images.githubusercontent.com/1711126/52967440-14d36080-3378-11e9-9250-3b6322917df2.png)

It does also make a big difference with tilesets, which often load in lots of images during camera fly throughs. 

Another test is on the Melbourne 3D Tileset, during a fly through, we see the following before (top) and after:

![image_bitmap_melbourne](https://user-images.githubusercontent.com/1711126/52967646-b5298500-3378-11e9-9512-52b5dbfbf409.png)

cesium-concierge commented:
Thanks for the pull request @OmarShehata!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/image-bitmap/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

mramato commented on Source/Core/FeatureDetection.js:
```diff
@@ -255,6 +253,71 @@ define([
+        }
+
+        var imageDataUri = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWP4////fwAJ+wP9CNHoHgAAAABJRU5ErkJggg==';
+        fetch(imageDataUri)
```
Why is this using the `fetch` API instead of `Resource.fetchBlob`?  Same question for where `fetch` is used in one spot in `Resource.js`.  In general we don't use the `fetch` API in Cesium because of lacking IE11 support.  That can/will change eventually, but is there a good reason for it here?

OmarShehata replied:
I needed a `Blob` because `createImageBitmap` blocks the main thread if it's given an `HTMLImageElement`. But if I can do that without `fetch` then I don't think I need it. I'll switch that and confirm that the behavior hasn't changed. Thanks for pointing this out.

mramato commented on Source/Core/Resource.js:
```diff
@@ -336,6 +339,14 @@ define([
+         *
+         * @type {Boolean}
+         */
+        this.flipImage = defaultValue(options.flipImage, true);
```
This feels like a leaky abstraction, I'm also not sure why it would go on the Resource constructor and not just on `fetchImage` (since that's the only place it applies, right?

mramato replied:
I also couldn't find from your PR description what this option is used for and why.  Isn't it a completely internal implementation detail?  Couldn't we always/never flip based on what we detect regarding createImageBitmap?  Why does the calling code need control over this?

OmarShehata replied:
Sorry for not making my justification here clearer.

We can't make decisions about whether to flip inside `Resource` because whether we flip or not depends on what we're going to do with this image. All textures in CesiumJS are flipped by default during texture upload. This seems to be because imagery expects this orientation.

Models with textures will disable this flipping to get the correct orientation. This means that there's no one correct orientation. If we always flip, models will be upside down. If we never flip, imagery will be upside down. I spent some time trying to see if I could get all imagery to be flipped in the shader so we can just simply never have to worry about flipping during texture upload but I couldn't get it to work correctly with the projection. 

As for why it's not just on `fetchImage`, you mean so that it would be like this?

```
Resource.prototype.fetchImage = function (preferBlob, flipY) {
```

The reason I implemented it the way it is right now is to get it to work for this case:

https://github.com/AnalyticalGraphicsInc/cesium/blob/e31eae334a4f8ddafe520ff59c0d2bf2ba16a33f/Source/Core/Resource.js#L899-L904

Where the local file-scope `fetchImage` function only takes a `resource` as a parameter. But I could just pass the flag there too. I do agree it's weird to have an option on `Resource` that only applies during a specific function call.

mramato replied:
Thanks for the details. Yes, I would just have fetchImage take an options object with `preferBlob` and `flipY` propreties (should be an easy deprecation since you can check if a bool was previously being passed).

OmarShehata commented:
Ok @mramato I've refactored it to use `Resource.fetchBlob` and remove `flipImage` from `Resource`. I've updated the PR description to document why we expose `flipY` on `fetchImage`. 

The only outstanding issue is https://github.com/AnalyticalGraphicsInc/cesium/pull/7579#discussion_r257797919. I'm still using the fetch API in `FeatureDetection` to avoid a circular dependency. I started changing it to take `Resource` as an argument but realized it was a lot simpler to just using fetch there since it's supported wherever createImageBitmap is. Otherwise I'd have to include `Resource` in a bunch of new places that don't need it just to pass it to FeatureDetection. I'm open to suggestions here.

mramato commented:
> Otherwise I'd have to include `Resource` in a bunch of new places that don't need it just to pass it to FeatureDetection. I'm open to suggestions here.

One idea would be to move the feature detection code into `Resource` instead of having it on FeatureDetection.  Since all image loads are async anyway, this also means you can check for support in one spot the first time an image is loaded.  It might actually simplify the code further.  If you don't like this idea, then using fetch might be okay (would need to look into a little more), but I kind of like the idea of putting the feature check closer to the use case anyway.

OmarShehata commented:
Ok, the good news is that the new approach works and not only simplifies the code but things are functionally better too. Here's some highlights:

* There will no longer be some images that'll slip through and load using the old code-path just because they were requested before `supportsImageBitmapOptions` finishes. It'll just wait for that promise to resolve.
  * So we no longer have to kick this off in Viewer.js. 
  * This also means we no longer have to wait for `supportsImageBitmapOptions` in the beforeAll of most specs! That's automatically handled now.
* I discovered if `prefersBlob` was true, it would fetch a blob, create a URL, then fetch that again, and create and ImageBitmap. This can now be avoided. 
* I added a deprecationWarning and discovered some areas in the code that I updated to the new function signature.

The other good news is that I found out why there was an extra parameter being passed to `fetchImage` that wasn't being used. `IonResource.fetchImage` has a `crossOriginAllowed` parameter. To fix this I either had to make `fetchImage` pass it down to here:

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/Resource.js#L911-L916

Or just remove that parameter from `IonResource`. I picked the latter since it was never working before for however long and this way it's consistent with `Resource`'s fetchImage. Let me know if you think this wasn't the right choice. 

I will bump this when CI passes.

OmarShehata commented:
@mramato this is good to look at. See https://github.com/AnalyticalGraphicsInc/cesium/pull/7579#issuecomment-465338985 for a summary of the changes.

mramato commented:
Nice, thanks again @OmarShehata. I'll do another pass on this either later tonight or first thing tomorrow.

mramato commented on Specs/Scene/TextureAtlasSpec.js:
```diff
@@ -274,10 +283,10 @@ defineSuite([
-        promises.push(atlas.addImage(blueImage.src, blueImage));
-        promises.push(atlas.addImage(bigRedImage.src, bigRedImage));
-        promises.push(atlas.addImage(bigBlueImage.src, bigBlueImage));
+        promises.push(atlas.addImage(guidArray[0], greenImage));
```
What were the changes in this file fore, this is much more obtuse than the previous code.  Assuming the change was necessary, having 4 variables (greenGuid, etc..) would make the tests much more readable.  These indexes amount to magic numbers otherwise.

OmarShehata replied:
This is necessary because `ImageBitmap` has no `src` property. This way it works regardless of whether the image is `Image` or `ImageBitmap`. 

You're right that just having `greenGuid` etc. variables looks much more readable.

mramato commented on Specs/Scene/DiscardMissingTileImagePolicySpec.js:
```diff
@@ -62,7 +70,12 @@ defineSuite([
+                } else {
+                    expect(Resource._Implementations.createImage).toHaveBeenCalled();
+                }
+
```
Minor nit, blank line.

mramato commented on Source/Core/FeatureDetection.js:
```diff
@@ -255,6 +253,15 @@ define([
+    var supportsCreateImageBitmapResult;
+    function supportsCreateImageBitmap() {
+        if (!defined(supportsCreateImageBitmapResult)) {
+            supportsCreateImageBitmapResult = defined(window.createImageBitmap);
```
In general avoid using `window.` for globals because window isn't always the global object (for example, Node.js and workers).  You can just use `createImageBitmap`.  Since it's such a trivial check, there's no need to cache a value here, this whole function ca be simplified to:

```
    function supportsCreateImageBitmap() {
        return typeof createImageBitmap === 'function';
    }
```

mramato commented on Source/Core/IonResource.js:
```diff
@@ -164,8 +164,13 @@ define([
-        return Resource.prototype.fetchImage.call(this, this._isExternal ? preferBlob : true, allowCrossOrigin);
+    IonResource.prototype.fetchImage = function (options) {
+        if (!this._isExternal) {
+            options = defaultValue(options, {});
```
Never use `defaultValue` when the second parameter will allocate an object.  `defaultValue` doesn't allow for short-circuiting so what you're really doing here is creating a new object and throwing it away.

That being said, we should never modify `options` objects passed in from the user anyway, so this function should always create a new options object and set values from the one passed in (if defined).

OmarShehata replied:
That's surprising to me because it seems to common in our code. A search for `defaultValue\([\w]+, {}` shows 20 occurrences (such as in `EllipsoidTerrainProvider.js` and `ImageryLayer.js`). Should these be fixed too?

mramato replied:
Forgotten or missed, write up a new issue and we'll make it as "good first issue" so new contributors can get their feet wet.

mramato commented on Source/Core/Resource.js:
```diff
@@ -374,6 +376,49 @@ define([
+            return supportsImageBitmapOptionsPromise.promise;
+        }
+
+        supportsImageBitmapOptionsPromise = when.defer();
```
I'm confused as to why we need a deferred promise in this function, can't we just cache the promise chain created with `Resource.fetchBlob` below?

mramato commented on Source/Core/Resource.js:
```diff
@@ -374,6 +376,49 @@ define([
+            url : imageDataUri
+        })
+            .then(function(blob) {
+                return when(createImageBitmap(blob, {
```
There's no need to wrap these calls in `when` if you are already inside of an existing when-based promise chain.

OmarShehata replied:
`createImageBitmap` returns a native promise, not a `when` promise. I do this because catching errors for native promises is `.catch` instead of `.otherwise`.

mramato replied:
Yes, I understand why you are doing it, but it's not necessary because you are already inside of a when based promise.  Looks at the console output of this [Sandcastle demo](https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=jZLLbtswEEV/ZaCNZCAgkW2iGq29CAIUDZCsCmgRihxZ04xIgY8obpF/L2W5tQtnUa5IDu+58+AWA6VBPGJwyWsUHUbdb9i11a/GwtlKnuEGyj7GMdxIqQ+6H0E4v5Pbw0FuErGRX8YxyCdljVYhMkqyBt9EHwcuT8T3VWNF7NFWXbI6krNVm01XcHTVzgbHKNjtqvLORZijV6AVM9kdaI8q4v2gdrihOKixXN0uQim/ObCIBqKDyasRWtQqBYQJQXkEsoFM3sLzlO2fYU7iwAXFmWr2YgF5jMlbmHL6bhIXhtWS0HmTaI4+eEIb1VxR7lbHNH4/lv0+p/hB3XSCflz+hTeElNsfyr9Al4l+ooAn6v+iOkWcPB5Zt8VVUYe4Z1wv6s80jM7HefqVEDLiMHImBNkm/YJR6BBmZS3/iGpDr0DmU1MsP2TrcjfIom8K0KxCyJEuMT/RT2yKdS3z+39k7JTJI354Rc9qPz/pr9dfl0shRC3z8VIVneNW+TPibw) showing what I mean.

Ultimately these wraps aren't needed, you only need to wrap it when it's a "top level" promise.

mramato commented on Source/Core/Resource.js:
```diff
@@ -827,10 +872,13 @@ define([
+     * @param {String|Object} options An object with the following properties.
+     * @param {Boolean} [options.preferBlob=false] If true, we will load the image via a blob.
+     * @param {Boolean} [options.flipY=true] If true, image will be vertially flipped during decode. Only applies if the browser supports `createImageBitmap`.
+     * @returns {Promise.<Image>|Promise.<ImageBitmap>|undefined} a promise that will resolve to the requested data when loaded. Returns undefined if <code>request.throttle</code> is true and the request does not have high enough priority.
```
What is the functional difference between `Image` and `ImageBitmap`?  Is this technically a breaking change but doesn't matter in practice?  Or will this break user code?  If all existing code will continue to work, great; but if it's a breaking change then we should note that.

For example, I just found out about https://developer.mozilla.org/en-US/docs/Web/API/ImageBitmap/close, it doesn't look like we are calling it anywhere.  Do we need to handle manually releasing the ImageBitmap?

OmarShehata replied:
This is technically a breaking change, yes. An example of this is actually our own code in `TextureAtlasSpec`. While you can do the same things with `Image` and `ImageBitmap`, there are some minor differences, like `ImageBitmap` doesn't have a `src` property. 

I contemplated adding an option to say `disableImageBitmap` or something but I can't think of any good reason why you'd want this in the real world. I will update `CHANGES.md` to mention it's a breaking change. 

As for `ImageBitmap.close`, I just did some quick research on this and can confirm that the memory is released when it's garbage collected. Looks like calling `close()` will allow us to explicitly control when it's released. However this will break cases where one source image is used to create multiple textures. Right now this mostly breaks specs. But I haven't been able to show that this would have a noticeable enough impact to add in this PR right now.

mramato commented on Source/Core/Resource.js:
```diff
@@ -827,10 +872,13 @@ define([
      *
-     * @param {Boolean} [preferBlob = false]  If true, we will load the image via a blob.
-     * @returns {Promise.<Image>|undefined} a promise that will resolve to the requested data when loaded. Returns undefined if <code>request.throttle</code> is true and the request does not have high enough priority.
+     * @param {String|Object} options An object with the following properties.
```
`options` itself needs to be marked optional, right?

Also, why is this `String|Object` instead of just `Object`?

mramato commented on Source/Core/Resource.js:
```diff
@@ -849,8 +897,16 @@ define([
-    Resource.prototype.fetchImage = function (preferBlob) {
-        preferBlob = defaultValue(preferBlob, false);
+    Resource.prototype.fetchImage = function (options) {
+        if (defined(options) && typeof options === 'boolean') {
```
No need for the defined check here, just the typeof check.

mramato commented on Source/Core/Resource.js:
```diff
@@ -849,8 +897,16 @@ define([
+                preferBlob : options
+            };
+        }
+        options = defaultValue(options, {});
```
Use `defaultValue.EMPTY_OBJECT` here instead of `{}` (as discussed above this avoids the allocation and lack of short-circuiting.

mramato commented on Source/Core/Resource.js:
```diff
@@ -860,28 +916,38 @@ define([
+                .then(function(blob) {
+                    var deferred = when.defer();
+                    Resource._Implementations.createImageBitmapFromBlob(blob, deferred, flipY);
+                    return deferred;
```
I don't understand why we need to pass a deferred to `createImageBitmapFromBlob`, it looks like that function is promise based anyway and should just return the promise it creates.

OmarShehata replied:
That's only to make it work with the `Resource._Implementations.createImage` which already has a deferred given to it. Although I guess if it needs to resolve a given deferred it can do:

```
Resource._Implementations.createImageBitmapFromBlob()
  .then(deferred.resolve)
  .otherwise(deferred.reject)
```

?

mramato replied:
Yes, I would push the deferred usage up to where it needs it.

mramato commented on Source/Core/Resource.js:
```diff
@@ -1755,26 +1825,50 @@ define([
+
+            image.src = url;
+        } else {
+            // createImageBitmap will block the main thread during decode if a blob is not supplied.
```
Is this true for all browsers?  Do you have a citation that states this to be the case?  Please add the url to the comment so future readers can learn more.

mramato replied:
@OmarShehata did I miss an answer to this?

mramato commented on Source/Core/Resource.js:
```diff
@@ -1755,26 +1825,50 @@ define([
+    };
 
-        image.src = url;
+    Resource._Implementations.createImageBitmapFromBlob = function(blob, deferred, flipY) {
```
Already mentioned this above, but there's no reason to take a deferred here, just return the promise you create below.  In general, avoid deferred.

mramato commented on Source/Core/loadImageFromTypedArray.js:
```diff
@@ -21,6 +31,21 @@ define([
+                        }));
+                    }
+
+                    return when(createImageBitmap(blob));
```
No need for these `when` wrappers.

OmarShehata replied:
I believe this is necessary. See https://github.com/AnalyticalGraphicsInc/cesium/pull/7579#discussion_r258934322

mramato commented on Source/Scene/DiscardMissingTileImagePolicy.js:
```diff
@@ -89,7 +89,9 @@ define([
         }
 
-        when(resource.fetchImage(true), success, failure);
+        when(resource.fetchImage({
```
I know this was existing code, but it's bad practice to use the `(promise, success, failure)` form of when.  Exceptions thrown in `promise` may not get properly handled.  Instead use `promise.then(success).otherwise(failure)` instead.

mramato commented:
This is really close @OmarShehata, hopefully that's my last round of feedback.

mramato commented:
Tested IE 11 and Edge just to be sure, both work as expected.

OmarShehata commented:
Ok! @mramato it should all be taken care of now. I did a final test to make sure it all still works on Chrome, Firefox, IE11 and Edge. I noticed some additional documentation that needed to be updated (that both fetchImage's now return `ImageBitmap`). 

Thanks for all the feedback and great suggestions. Let me know if you notice anything else and if CHANGES.md looks good.

mramato commented:
Last question, what happens if createImageBitmapOptions are not supported and someone passes in `flipY: true`?  Is it just ignored or do we generate a error.  At first blush, I would expect a `RuntimeError` but I could be convinced `DeveloperError` would be better.

OmarShehata commented:
@mramato it is simply ignored. I am a firm believer that silent errors are evil, so this was a difficult design decision. Let me explain some context.

The behavior we expect is best summed up by this test: 

https://github.com/AnalyticalGraphicsInc/cesium/blob/bf46f89eefb9055f9ae60575ce324345f63ffa8b/Specs/Renderer/TextureSpec.js#L199-L208

We expect that it is only possible to flip a texture if `ImageBitmapOptions` are not supported, and vice versa. Here's the table of the behaviors:

|   Can flip during           | Firefox            | Chrome             | Other              |
|--------------|--------------------|--------------------|--------------------|
| fetch    | :x:                | :heavy_check_mark: | :x:                |
| texture upload | :heavy_check_mark: | :x:                | :heavy_check_mark: |

So if you want to make sure something is flipped,  just flip in both steps. If one day a browser supports flipping on both decode and on upload then imagery will be upside down. However, I think this is a safe assumption because the specification does state that flipY should not apply to ImageBitmap (emphasis mine):

> See Pixel Storage Parameters for WebGL-specific pixel storage parameters that affect the behavior of this function when it is called with **any argument type other than ImageBitmap**. 
Source: https://www.khronos.org/registry/webgl/specs/latest/1.0/#TEXIMAGE2D_HTML

Perhaps what I need to do now is add a check during texture upload of any imagery, to not bother flipping if `ImageBitmapOptions` is supported, to mitigate against this possibility. I didn't do this because creating textures is synchronous and adding that promise check everywhere there might be difficult. 

### Why not just flip in one area?

We contemplated just removing `flipY` from texture uploads, and have it automatically figure out, if an image was fetched with `flipY`, even though `ImageBitmapOptions` isn't supported, it should store this meta-information and flip during texture upload. But not all textures are created from source images, and not all requested images get uploaded as textures.

OmarShehata commented:
I'm open to ideas here.

mramato commented:
Thanks for the details, as discussed offline, for Firefox to work, this PR relies on the fact that these two Firefox bugs offset each other:

* https://bugzilla.mozilla.org/show_bug.cgi?id=1367251
* https://bugzilla.mozilla.org/show_bug.cgi?id=1335594

If they fix one of those without the other, all imagery in Cesium will become upside down (as of this PR).

The options are:

1. Don't worry about it, deal with it if/when those issues get fixed. The downside is a new Firefox release could cause upside-down images in all existing CesiumJS applications (seems like a big risk to take).
2. Don't use ImageBitmap on Firefox at all until the issues are addressed (safest bet but does not give any performance benefit to Firefox in the meantime).  We could certainly reach out to the Firefox team and try to see if we can get someone on it.
3. Add lots of special cases throughout the Cesium code base to workaround the broken bug. (probably way to messy).

@OmarShehata please correct me if any of this is wrong

We could use a second opinion here CC @pjcozzi @lilleyse

mramato commented:
I think I lean towards 2, above.  I think 1 is too big a risk.

That being said, if 1 ever becomes a problem, we could tell users how to monkey-patch `supportsCreateImageBitmap` to fix old Cesium versions, so it all depends on our risk tolerance vs making Firefox less janky in the meantime.

OmarShehata commented:
My vote is on 1, in that I'm betting on either Firefox will not fix either or fix both together eventually. If people fixing older applications is as easy as `Cesium.FeatureDetection.supportsCreateImageBitmap = function(){ return false };` I think it's worth the risk.

mramato commented:
Firefox is the last sticking for this PR, as soon as enough people chime in and we come to a decision, I'm okay with merging.

lilleyse commented:
Both issues are surprisingly old. I think it's slightly better to go with 2 since this feature is helpful but not essential to get into Firefox right away.

mramato commented:
What's the status of this PR?  Last I heard we were trying to get some info/advisement from the Firefox team?

Do we want to get this into the Friday release?

OmarShehata commented:
> What's the status of this PR? Last I heard we were trying to get some info/advisement from the Firefox team?

I still haven't heard back. Perhaps this might be better to merge after release at this point?

mramato commented:
Merging after is fine.  I would re-recommend doing 2 and then making additional changes once we hear back from the FF team.  Just bump this when it's ready.  Thanks!

OmarShehata commented:
We didn't get a definitive answer back, but it looks like there are no known plans for those bugs. The consensus in this [ThreeJS](https://github.com/mrdoob/three.js/issues/11746#issuecomment-414831562) thread is to just disable for Firefox. We're already detecting it by actually calling it to see if it works, so no need to detect and disable for any specific browser. 

I've updated the code to only use createImageBitmap when ImageBitmapOptions is supported. This does simplify the calling code a bit since we no longer need to account for the two paths.

@mramato no rush on this review at the moment, but would be good to get this far-reaching in sometime this/next week.

mramato commented on Source/Core/FeatureDetection.js:
```diff
@@ -269,6 +267,10 @@ define([
         }
     }
 
+    function supportsCreateImageBitmap() {
```
I don't think this or `supportsImageBitmapOptionsSync` is really needed you can delete them and simplify the code.  You should just be able to use `Resource.supportsImageBitmapOptions` everywhere we need to know, I did a quick look into this locally and everywhere this would be used is already promise-based.

mramato commented on Source/Core/Resource.js:
```diff
@@ -1775,6 +1860,47 @@ define([
+                    return;
+                }
+
+                Resource.fetchBlob({
```
This should have a return statement and then you should move all of the following `then` calls up a level.

OmarShehata replied:
Can you explain why? I'm assuming this is what you mean:

```
Resource.supportsImageBitmapOptions()
    .then(function(result) {
        if (!result) {
            loadImageElement(url, crossOrigin, deferred);
            return;
        }

        return Resource.fetchBlob({
            url: url
        });
    })
    .then(function(blob) {
        return Resource._Implementations.createImageBitmapFromBlob(blob, flipY);
    })
    .then(deferred.resolve).otherwise(deferred.reject);
```

This will run the `.then(function(blob)` anyway even if we `loadImageElement` and don't fetch the blob, and then `blob` will be `undefined`. 

I can just move the `.then(deferred.resolve).otherwise(deferred.reject);` outside and I that should work fine.

OmarShehata replied:
I can see `Resource.prototype.fetchImage` checks if the result is defined at each step in the promise chain to make sure it doesn't run when it terminates early like that. I can follow that pattern.

mramato replied:
Because if for some reason `supportsImageBitmapOptions` rejected, you would  never call `deferred.reject` (or `deferred.resolve`).

mramato replied:
> I can see `Resource.prototype.fetchImage` checks if the result is defined at each step in the promise chain to make sure it doesn't run when it terminates early like that. I can follow that pattern.

I'm not sure what you mean by this, please stop by if you still have questions.

OmarShehata replied:
Thanks for the explanation. I'm pretty sure I've got it locked down now. What I meant was, Resource has a similar promise chain when fetching an image. When we return undefined, all the subsequent then's are still going to run. So you need to add a check at each subsequent then() that the result is defined. 

Here is where Resource is doing this: 

https://github.com/AnalyticalGraphicsInc/cesium/blob/8bc364ce5eca3c42d84fe21386b0d5a50be82c2d/Source/Core/Resource.js#L886-L889

I just followed a similar pattern in `createImage`:

https://github.com/AnalyticalGraphicsInc/cesium/blob/a6324e2bcf8b3547b4adb589e7ee789321997678/Source/Core/Resource.js#L1855-L1882

Notice that `deferred.resolve` is only called in the ImageBitmap codepath. Since `loadImageElement` calls it on the `Image` on `onload`.

mramato replied:
I think this whole file could probably use some clean-up.  Some of the (old) if checks make reading the code pretty confusing.  No action on your part, and this is nothing you did; just mentioning it because I may open a small clean-up PR to take a pass (after merging this).

OmarShehata commented:
@mramato all good now. Last commit makes all testing rely just on the `Resource.supportsImageBitmapOptions` promise. Tested again on Chrome, Edge, Firefox and IE11.

mramato commented on Specs/isImageOrImageBitmap.js:
```diff
@@ -0,0 +1,22 @@
+        Resource) {
+    'use strict';
+
+    function isImageOrImageBitmap(image) {
```
This would be better as a custom Jasmine matcher than a standlone function, I'll just submit a fix to master after merging.

mramato commented:
Thanks @OmarShehata!  I think we can finally call this done.  Hopefully Firefox and Safari will get in line and add full support so we can benefit there as well.
